simulate/simulate.cc

```diff
@@ -0,0 +1,1834 @@
+          mju_mulMatTVec(simulate->pert.localpos, simulate->d->xmat+9*simulate->pert.select, tmp, 3, 3);
+        } else {
+          simulate->pert.select = 0;
+          simulate->pert.skinselect = -1;
+        }
+      }
+
+      // stop perturbation on select
+      simulate->pert.active = 0;
+    }
+
+    return;
+  }
+
+  // 3D release
+  if (state->type==mjEVENT_RELEASE && state->dragrect==3 && simulate->m) {
+    // stop perturbation
+    simulate->pert.active = 0;
+
+    return;
+  }
+
+  // 3D move
+  if (state->type==mjEVENT_MOVE && state->dragrect==3 && simulate->m) {
+    // determine action based on mouse button
+    mjtMouse action;
+    if (state->right) {
+      action = state->shift ? mjMOUSE_MOVE_H : mjMOUSE_MOVE_V;
+    } else if (state->left) {
+      action = state->shift ? mjMOUSE_ROTATE_H : mjMOUSE_ROTATE_V;
+    } else {
+      action = mjMOUSE_ZOOM;
+    }
+
+    // move perturb or camera
+    mjrRect r = state->rect[3];
+    if (simulate->pert.active)
+      mjv_movePerturb(simulate->m, simulate->d, action, state->dx/r.height, -state->dy/r.height,
+                      &simulate->scn, &simulate->pert);
+    else
+      mjv_moveCamera(simulate->m, action, state->dx/r.height, -state->dy/r.height,
+                     &simulate->scn, &simulate->cam);
+
+    return;
+  }
+}
+
+void uiRender(mjuiState* state) {
+  mj::Simulate* simulate = (mj::Simulate*)(state->userdata);
+  simulate->render();
+}
+
+// drop file callback
+void drop(mj::Simulate* simulate, int count, const char** paths) {
+  // make sure list is non-empty
+  if (count>0) {
+    mju::strcpy_arr(simulate->filename, paths[0]);
+    simulate->loadrequest = 1;
+  }
+}
+
+void uiDrop(mjuiState* state, int count, const char** paths) {
+  mj::Simulate* simulate = (mj::Simulate*)(state->userdata);
+  drop(simulate, count, paths);
+}
+
+} // end unnamed namespace
+
+namespace mujoco {
+namespace mju = ::mujoco::sample_util;
+
+//---------------------------------- init -------------------------------------------------
+
+// create object and initialize the simulate ui
+Simulate::Simulate(void) {
+  // init GLFW, set timer callback (milliseconds)
+  if (!glfwInit()) {
+    mju_error("could not initialize GLFW");
+  }
+  mjcb_time = timer;
+
+  // multisampling
+  glfwWindowHint(GLFW_SAMPLES, 4);
+  glfwWindowHint(GLFW_VISIBLE, 1);
+
+  // get videomode and save
+  this->vmode = *glfwGetVideoMode(glfwGetPrimaryMonitor());
+
+  // create window
+  this->window = glfwCreateWindow((2*this->vmode.width)/3, (2*this->vmode.height)/3,
+                            "Simulate", nullptr, nullptr);
+  if (!this->window) {
+    glfwTerminate();
+    mju_error("could not create window");
+  }
+
+  // save window position and size
+  glfwGetWindowPos(this->window, this->windowpos, this->windowpos+1);
+  glfwGetWindowSize(this->window, this->windowsize, this->windowsize+1);
+
+  // make context current, set v-sync
+  glfwMakeContextCurrent(this->window);
```

Comment:
Since the `Simulate` is supposed to encapsulate the entire window and rendering context, it really should create and maintain its own thread. As things stand, this class cannot be safely use alongside any other GLFW (or indeed OpenGL) code.

Reply:
Good catch, my inexperience with GLFW and OpenGL is showing.

I'll refactor the rendering thread currently run from `main()` so that it is a part of `Simulate`. The incorrectly named `simulate_thread` in `main.cc` will be run from `main()` instead.

`Simulate`'s constructor will be modified to accept a mutex that will be used by the thread it launches to control access to the model and data structures.

The interaction between the rendering thread, the physics thread, and the `ctrlnoise` variable that occurs when a model xml file is drag and dropped will be reworked as well.

Reply:
I don't think it'll be appropriate to launch the thread in the object's constructor. A better way to do it would be to create the window in the constructor, and have methods for launching and stopping the encapsulated thread.

About the mutex: why not have the Simulate class own it?

Reply:
Yes I agree using methods to start/stop `Simulate`'s thread is better than launching the thread from the constructor. While there are already things that can fail in the constructor, the thread would be too much.

I think the user of `Simulate` should own the mutex to because multiple `Simulate` instances should be able to launched simultaneously on the same `model` and `data` instances, and then it doesn't seem to make sense for `Simulate` to contain the mutex. Additionally, I can envision other objects with purpose like `Simulate` (maybe logger or simulated camera rendering pool) that would also require exclusive access to the model and data using the same mutex.

Because of this I also think the model and data should be arguments to `Simulate`'s constructor.

The creation of the model and data struct's will need to be reworked to be outside of `Simulate`'s direct control in order for this to make sense, but I think it is worth it.

Reply:
Originally, I thought that reworking `Simulate` to support multiple `Simulate`'s running on the same model and data should be a second PR since it requires a decent amount of change in the model loading code and seems to be more than a refactoring to me. But maybe it is more appropriate to just do that in this PR.

Reply:
@yuvaltassa pointed out that there might be some limitation on which thread is allowed to make rendering calls on macOS. It _might_ be necessary to create the window on the same thread on which you intend to make rendering calls.

I don't suppose you're able to check things on macOS? If that's the case, I can try it out myself once you have it working on Linux to confirm.

Reply:
I was worried about something like that popping up. Would a fix be to make the window related variables private (everything will need to be made private anyway), and let `Simulate`'s thread create the window instead of the constructor?

This seems like it would be fine given that if `Simulate`'s creates its own thread the interface is going to change significantly to just be:

* the constructor
* the start thread function
* the stop thread function
* threadsafe getters of key `Simulate` variables that users may be interested in (perturbations, exitrequest, etc)

Reply:
To answer the question, no I don't have a macOS to check things (which is why there is no macOS Makefile in this PR either...)